src/bidiMapper/domains/input/InputSource.ts

```diff
@@ -0,0 +1,153 @@
+/**
+ * Copyright 2023 Google LLC.
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Input} from '../../../protocol/protocol.js';
+
+export import SourceType = Input.SourceActionsType;
+
+export class NoneSource {
+  type = SourceType.None as const;
+}
+export class KeySource {
+  type = SourceType.Key as const;
+  pressed = new Set<string>();
+
+  #modifiers = 0;
+  get modifiers(): number {
+    return this.#modifiers;
+  }
+  get alt(): boolean {
+    return (this.#modifiers & 1) === 1;
+  }
+  set alt(value: boolean) {
+    this.#setModifier(value, 1);
+  }
+  get ctrl(): boolean {
+    return (this.#modifiers & 2) === 2;
+  }
+  set ctrl(value: boolean) {
+    this.#setModifier(value, 2);
+  }
+  get meta(): boolean {
+    return (this.#modifiers & 4) === 4;
+  }
+  set meta(value: boolean) {
+    this.#setModifier(value, 4);
+  }
+  get shift(): boolean {
+    return (this.#modifiers & 8) === 8;
+  }
+  set shift(value: boolean) {
+    this.#setModifier(value, 8);
+  }
+
+  #setModifier(value: boolean, bit: number) {
+    if (value) {
+      this.#modifiers |= bit;
+    } else {
+      this.#modifiers ^= bit;
+    }
+  }
+}
+
+interface ClickContext {
+  x: number;
+  y: number;
+  timeStamp: number;
+}
+
+export class PointerSource {
+  type = SourceType.Pointer as const;
+  subtype: Input.PointerType;
+  pointerId: number;
+  pressed = new Set<number>();
+  x = 0;
+  y = 0;
+
+  constructor(id: number, subtype: Input.PointerType) {
+    this.pointerId = id;
+    this.subtype = subtype;
+  }
+
+  get buttons(): number {
+    let buttons = 0;
+    for (const button of this.pressed) {
+      switch (button) {
+        case 0:
+          buttons |= 1;
+          break;
+        case 1:
+          buttons |= 4;
+          break;
+        case 2:
+          buttons |= 2;
+          break;
+        case 3:
+          buttons |= 8;
+          break;
+        case 4:
+          buttons |= 16;
+          break;
+      }
+    }
+    return buttons;
+  }
+
+  // --- Platform-specific state starts here ---
+  // This code should match https://source.chromium.org/chromium/chromium/src/+/refs/heads/main:ui/events/event.cc;l=479
```

Comment:
Is it because of CDP considers a double click as one click, not as 2 separate clicks?

Reply:
It's better to say CDP doesn't know what a double click is, but knows about a click count parameter that can trigger double clicks if it puts `2` in it.

Reply:
Could you please add a comment about that in the code?